sql/migrate/migrate.go

```diff
@@ -373,6 +397,117 @@ type templateFile struct {
 // Name implements the File interface.
 func (f *templateFile) Name() string { return f.n }
 
+// Filename of the migration directory integrity sum file.
+const hashFile = "atlas.sum"
+
+// HashFile represents the integrity sum file of the migration dir.
+type HashFile []struct{ N, H string }
+
+// HashSum reads the whole dir, sorts the files by name and creates a HashSum from its contents.
+func HashSum(dir Dir) (HashFile, error) {
+	var hs HashFile
+	h := sha256.New()
+	err := fs.WalkDir(dir, "", func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		// If this is the integrity sum file do not include it into the sum.
+		if filepath.Base(path) == hashFile {
+			return nil
+		}
+		if !d.IsDir() {
+			f, err := dir.Open(path)
+			if err != nil {
+				return err
+			}
+			defer f.Close()
+			if _, err = io.Copy(h, f); err != nil {
+				return err
+			}
+			hs = append(hs, struct{ N, H string }{path, base64.StdEncoding.EncodeToString(h.Sum(nil))})
+			h.Reset()
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return hs, nil
+}
+
+// ErrChecksumMismatch is returned from Validate if the hash sums don't match.
+var ErrChecksumMismatch = errors.New("checksum mismatch")
+
+// Validate checks if the migration dir is in sync with its sum file.
+// If they don't match ErrChecksumMismatch is returned.
+func Validate(dir Dir) error {
+	f, err := dir.Open(hashFile)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	b, err := io.ReadAll(f)
+	if err != nil {
+		return err
+	}
+	var fh HashFile
+	if err := fh.UnmarshalText(b); err != nil {
+		return nil
+	}
+	mh, err := HashSum(dir)
+	if err != nil {
+		return err
+	}
+	if !fh.equals(mh) {
+		return ErrChecksumMismatch
+	}
+	return nil
+}
+
+// MarshalText implements encoding.TextMarshaler.
+func (s HashFile) MarshalText() ([]byte, error) {
+	buf := new(bytes.Buffer)
+	sha := sha256.New()
+	for _, f := range s {
+		buf.WriteString(fmt.Sprintf("%s h:%s\n", f.N, f.H))
+		sha.Write([]byte(f.H))
+	}
+	ret := new(bytes.Buffer)
+	ret.WriteString(fmt.Sprintf("sum h:%s\n", base64.StdEncoding.EncodeToString(sha.Sum(nil))))
+	ret.Write(buf.Bytes())
+	return ret.Bytes(), nil
+}
+
+// UnmarshalText implements encoding.TextUnmarshaler.
+func (s *HashFile) UnmarshalText(b []byte) error {
+	sc := bufio.NewScanner(bytes.NewReader(b))
+	sc.Scan() // skip first line
+	for sc.Scan() {
+		li := strings.SplitN(sc.Text(), "h:", 2)
+		if len(li) != 2 {
+			return errors.New("invalid sum file")
+		}
+		*s = append(*s, struct{ N, H string }{strings.TrimSpace(li[0]), li[1]})
+	}
+	if err := sc.Err(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// equals checks if the o is equal to s.
+func (s HashFile) equals(o HashFile) bool {
+	if len(s) != len(o) {
+		return false
+	}
+	for i := range s {
+		if s[i].N != o[i].N || s[i].H != o[i].H {
+			return false
+		}
+	}
+	return true
+}
```

Comment:
Don't you need to compare also the `sum` value?

Reply:
No, it is only used for VCS merge conflict and not for the validity of the file.